src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/mappers/HaFlowHistoryMapper.java

```diff
@@ -0,0 +1,339 @@
+                .collect(Collectors.toList());
+    }
+
+    public abstract List<List<PathNodePayload>> persistenceToPayloadList(
+            List<List<HaFlowEventDump.PathNodePayload>> value);
+
+    @Mapping(target = "timestampIso",
+            expression = "java(haFlowEventAction.getTimestamp().atOffset(ZoneOffset.UTC).toString())")
+    public abstract HaFlowHistoryPayload toHaFlowHistoryPayload(HaFlowEventAction haFlowEventAction);
+
+    /**
+     * Creates an HA-flow dump that represents the same HA-flow data but structured in a different way.
+     * This representation is supposed to be used to store and pass around history events.
+     * @param haFlow the source of the data
+     * @param correlationId operation ID
+     * @param dumpType dump type
+     * @return dump of the HA-flow
+     */
+    public HaFlowDumpData toHaFlowDumpData(HaFlow haFlow, String correlationId, DumpType dumpType) {
+        if (haFlow == null) {
+            return HaFlowDumpData.empty();
+        }
+
+        return HaFlowDumpData.builder()
+                .dumpType(dumpType)
+                .taskId(correlationId)
+                .haFlowId(haFlow.getHaFlowId())
+                .affinityGroupId(haFlow.getAffinityGroupId())
+                .allocateProtectedPath(haFlow.isAllocateProtectedPath())
+                .description(haFlow.getDescription())
+                .diverseGroupId(haFlow.getDiverseGroupId())
+                .encapsulationType(haFlow.getEncapsulationType())
+                .flowTimeCreate(haFlow.getTimeCreate())
+                .flowTimeModify(haFlow.getTimeModify())
+                .ignoreBandwidth(haFlow.isIgnoreBandwidth())
+                .maxLatency(haFlow.getMaxLatency())
+                .maxLatencyTier2(haFlow.getMaxLatencyTier2())
+                .maximumBandwidth(haFlow.getMaximumBandwidth())
+                .pathComputationStrategy(haFlow.getPathComputationStrategy())
+                .periodicPings(haFlow.isPeriodicPings())
+                .pinned(haFlow.isPinned())
+                .priority(haFlow.getPriority())
+                .sharedInnerVlan(haFlow.getSharedInnerVlan())
+                .sharedOuterVlan(haFlow.getSharedOuterVlan())
+                .sharedPort(haFlow.getSharedPort())
+                .sharedSwitchId(haFlow.getSharedSwitchId())
+                .status(haFlow.getStatus())
+                .statusInfo(haFlow.getStatusInfo())
+                .strictBandwidth(haFlow.isStrictBandwidth())
+                .haSubFlows(toHaSubFlowDumps(haFlow))
+                .forwardPath(toHaFlowPathDump(haFlow.getForwardPath()))
+                .reversePath(toHaFlowPathDump(haFlow.getReversePath()))
+                .protectedForwardPath(toHaFlowPathDump(haFlow.getProtectedForwardPath()))
+                .protectedReversePath(toHaFlowPathDump(haFlow.getProtectedReversePath()))
+                .build();
+    }
+
+    private List<HaSubFlowDump> toHaSubFlowDumps(HaFlow haFlow) {
```

Comment:
@artHalMirantis here is that place that sets the HA flow and subflow IDs. Essentially, the history saves whatever is there in the HaFlow object. So if some ID is null in the HA flow object, then it is null in the history. So seems correct?